src/main/kotlin/org/jitsi/nlj/rtcp/KeyframeRequester.kt

```diff
@@ -31,39 +39,128 @@ import org.jitsi.rtp.rtcp.rtcpfb.payload_specific_fb.RtcpFbFirPacketBuilder
  * on what the client supports
  * 3) Aggregation.  This class will pace outgoing requests such that we don't spam the sender
  */
-class KeyframeRequester : ObserverNode("Keyframe Requester") {
+class KeyframeRequester : TransformerNode("Keyframe Requester") {
+
+    companion object {
+        private const val DEFAULT_WAIT_INTERVAL_MS = 100
+    }
+
+    var waitIntervalMs = DEFAULT_WAIT_INTERVAL_MS
+
+    override fun transform(packetInfo: PacketInfo): PacketInfo? {
+        val packet = packetInfo.packet
+        val pliOrFir = when {
+            packet is CompoundRtcpPacket -> {
+                packet.packets.first { it is RtcpFbPliPacket || it is RtcpFbFirPacket }
+            }
+            packet is RtcpFbFirPacket -> packet
+            packet is RtcpFbPliPacket -> packet
+            else -> null
+        }
+
+        var forward = true
+        when {
+            pliOrFir is RtcpFbPliPacket -> {
+                when {
+                    hasPliSupport -> {
+                        // Forward PLI
+                        forward = canSendKeyframeRequest(pliOrFir.mediaSenderSsrc, System.currentTimeMillis())
+                    }
+                    hasFirSupport -> {
+                        // Translate to FIR
+                        requestKeyframe(pliOrFir.mediaSenderSsrc)
+                        forward = false
+                    }
+                    else -> {
+                        // Can't do anything
+                        forward = false
+                    }
+                }
+            }
+            pliOrFir is RtcpFbFirPacket -> {
+                when {
+                    hasPliSupport -> {
+                        // Translate to PLI
+                        requestKeyframe(pliOrFir.mediaSenderSsrc)
+                        forward = false
+                    }
+                    hasFirSupport -> {
+                        // Forward FIR
+                        forward = canSendKeyframeRequest(pliOrFir.mediaSenderSsrc, System.currentTimeMillis())
+                        if (forward) {
+                            pliOrFir.seqNum = firCommandSequenceNumber.incrementAndGet()
+                        }
+                    }
+                    else -> {
+                        // Can't do anything
+                        forward = false
+                    }
+                }
+            }
+        }
+
+        return if (forward) packetInfo else null
+    }
+
     // Map a SSRC to the timestamp (in ms) of when we last requested a keyframe for it
     private val keyframeRequests = mutableMapOf<Long, Long>()
-    private var firCommandSequenceNumber: Int = 0
+    private var firCommandSequenceNumber: AtomicInteger = AtomicInteger(0)
```

Comment:
if this is an atomicint i think we can actually make it a val (since we never re-assign this directly, just inc/dec the atomicint)